cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/AbstractProgramRuntimeService.java

```diff
@@ -289,9 +289,19 @@ private ProgramOptions updateProgramOptions(ArtifactId artifactId, ProgramId pro
     // Build the system arguments
     ImmutableMap.Builder<String, String> builder = ImmutableMap.builder();
     builder.putAll(options.getArguments().asMap());
-    builder.putAll(getExtraProgramOptions());
-    builder.put(ProgramOptionConstants.RUN_ID, runId.getId());
-    builder.put(ProgramOptionConstants.ARTIFACT_ID, Joiner.on(':').join(artifactId.toIdParts()));
+    // don't add these system arguments if they're already there
+    // this can happen if this is a program within a workflow, and the workflow already added these arguments
+    for (Map.Entry<String, String> extraOption : getExtraProgramOptions().entrySet()) {
+      if (!options.getArguments().hasOption(extraOption.getKey())) {
+        builder.put(extraOption.getKey(), extraOption.getValue());
+      }
+    }
+    if (!options.getArguments().hasOption(ProgramOptionConstants.RUN_ID)) {
+      builder.put(ProgramOptionConstants.RUN_ID, runId.getId());
```

Comment:
Simpler if you just use normal `Map` and call `putIfAbsent` (available since Java 8).